File: core/rom_renamer.py
"""
OpenROM ROM Renamer — DAT-based ROM identification and renaming.
Supports No-Intro and Redump DAT files (same XML format).
M5 Dev | GPL v3
"""

import os
import shutil
import xml.etree.ElementTree as ET
from collections.abc import Callable
from dataclasses import dataclass

# ── DAT storage ───────────────────────────────────────────────────────────────

def get_dats_dir() -> str:
    """Returns the directory where OpenROM stores imported DAT files."""
    from core.config import get_config_dir
    dats_dir = os.path.join(get_config_dir(), "dats")
    os.makedirs(dats_dir, exist_ok=True)
    return dats_dir

def import_dat(dat_path: str) -> dict:
    """
    Import a DAT file into OpenROM's config directory.
    Returns info dict with keys: name, source, system, game_count, stored_path
    Raises ValueError if the file is not a valid DAT.
    """
    if not os.path.isfile(dat_path):
        raise ValueError(f"File not found: {dat_path}")

    tree = ET.parse(dat_path)
    root = tree.getroot()

    header = root.find("header")
    if header is None:
        raise ValueError("Invalid DAT file: no <header> element found")

    name_el = header.find("name")
    url_el  = header.find("url")

    dat_name = name_el.text.strip() if name_el is not None else os.path.basename(dat_path)
    dat_url  = url_el.text.strip()  if url_el  is not None else ""

    source = "No-Intro" if "no-intro" in dat_url.lower() else \
             "Redump"   if "redump"   in dat_url.lower() else "Unknown"

    game_count = len(root.findall("game"))

    # Store with sanitized name
    safe_name = "".join(c if c.isalnum() or c in " .-_()" else "_" for c in dat_name)
    stored_path = os.path.join(get_dats_dir(), f"{safe_name}.dat")
    shutil.copy2(dat_path, stored_path)

    return {
        "name":        dat_name,
        "source":      source,
        "system":      dat_name,
        "game_count":  game_count,
        "stored_path": stored_path,
        "url":         dat_url,
    }

def list_dats() -> list[dict]:
    """
    Returns a list of all imported DATs with their metadata.
    Each dict has: name, source, game_count, stored_path
    """
    dats_dir = get_dats_dir()
    result = []
    for fname in sorted(os.listdir(dats_dir)):
        if not fname.endswith(".dat"):
            continue
        fpath = os.path.join(dats_dir, fname)
        try:
            info = _read_dat_header(fpath)
            info["stored_path"] = fpath
            result.append(info)
        except Exception:
            pass
    return result

def remove_dat(stored_path: str) -> None:
    """Remove an imported DAT file."""
    if os.path.isfile(stored_path):
        os.remove(stored_path)

def _read_dat_header(dat_path: str) -> dict:
    """Read header info from a DAT file without loading all games."""
    tree = ET.parse(dat_path)
    root = tree.getroot()
    header  = root.find("header")
    name_el = header.find("name") if header is not None else None
    url_el  = header.find("url")  if header is not None else None
    dat_name = name_el.text.strip() if name_el is not None else os.path.basename(dat_path)
    dat_url  = url_el.text.strip()  if url_el  is not None else ""
    source   = "No-Intro" if "no-intro" in dat_url.lower() else \
               "Redump"   if "redump"   in dat_url.lower() else "Unknown"
    game_count = len(root.findall("game"))
    return {"name": dat_name, "source": source, "game_count": game_count}


# ── CRC32 calculation ─────────────────────────────────────────────────────────

def calc_crc32(filepath: str, on_progress: Callable[[float], None] | None = None) -> str:
    """
    Calculate CRC32 of a file. Returns lowercase hex string e.g. 'dbef116c'.
    Calls on_progress(percent) periodically if provided.
    """
    import zlib
    crc = 0
    file_size = os.path.getsize(filepath)
    bytes_read = 0
    chunk_size = 1024 * 1024  # 1 MB

    with open(filepath, "rb") as f:
        while True:
            chunk = f.read(chunk_size)
            if not chunk:
                break
            crc = zlib.crc32(chunk, crc)
            bytes_read += len(chunk)
            if on_progress and file_size > 0:
                on_progress(bytes_read / file_size * 100)

    return format(crc & 0xFFFFFFFF, "08x")


# ── DAT lookup ────────────────────────────────────────────────────────────────

def load_dat_index(dat_path: str) -> dict[str, dict]:
    """
    Load a DAT file and return a dict indexed by CRC32 (lowercase hex).
    Each value is: {name, description, rom_name, size, md5, sha1}
    """
    tree  = ET.parse(dat_path)
    root  = tree.getroot()
    index = {}

    for game in root.findall("game"):
        game_name = game.get("name", "")
        desc_el   = game.find("description")
        desc      = desc_el.text.strip() if desc_el is not None else game_name

        for rom in game.findall("rom"):
            crc = (rom.get("crc") or "").strip().lower().zfill(8)
            if not crc:
                continue
            index[crc] = {
                "name":        game_name,
                "description": desc,
                "rom_name":    rom.get("name", ""),
                "size":        int(rom.get("size") or 0),
                "md5":         (rom.get("md5")  or "").lower(),
                "sha1":        (rom.get("sha1") or "").lower(),
            }

    return index


# ── Scan results ──────────────────────────────────────────────────────────────

@dataclass
class RomScanResult:
    filepath:      str
    filename:      str
    crc32:         str          = ""
    matched:       bool         = False
    canonical_name: str         = ""   # game name from DAT
    rom_name:      str          = ""   # full rom filename from DAT
    dat_source:    str          = ""
    error:         str          = ""

    @property
    def suggested_filename(self) -> str:
        """Returns the canonical filename (rom_name from DAT) if matched."""
        if not self.matched or not self.rom_name:
            return self.filename
        return self.rom_name


# ── Main scan function ────────────────────────────────────────────────────────

def scan_folder(
    folder: str,
    dat_paths: list[str],
    on_progress: Callable[[str, float], None] | None = None,
    on_result:   Callable[[RomScanResult], None] | None = None,
) -> list[RomScanResult]:
    """
    Scan a folder of ROMs against one or more DAT files.

    Args:
        folder:      Path to folder containing ROM files
        dat_paths:   List of DAT file paths to match against
        on_progress: Callback(filename, percent) for progress updates
        on_result:   Callback(RomScanResult) called as each file finishes

    Returns:
        List of RomScanResult, one per ROM file scanned
    """
    if not os.path.isdir(folder):
        raise ValueError(f"Folder not found: {folder}")

    # Load all DAT indexes
    dat_indexes: list[tuple[dict, str, str]] = []  # (index, dat_name, source)
    skipped_dats: list[tuple[str, str]] = []

    for dp in dat_paths:
        try:
            header = _read_dat_header(dp)
            index  = load_dat_index(dp)
            dat_indexes.append((index, header["name"], header["source"]))
        except Exception as e:
            skipped_dats.append((os.path.basename(dp), str(e)))

    for dat_name, reason in skipped_dats:
        if on_progress:
            on_progress(f"[WARN] Skipped unreadable DAT: {dat_name} ({reason})", 0.0)

    # Collect ROM files
    extensions = {
        ".smc", ".sfc", ".nes", ".gba", ".gb", ".gbc",
        ".n64", ".z64", ".v64", ".ndd",
        ".md", ".gen", ".sms", ".gg", ".32x",
        ".pce", ".iso", ".bin", ".img", ".chd",
        ".ws", ".wsc", ".ngp", ".ngc",
        ".lnx", ".vb", ".vec", ".a26", ".a52", ".j64",
    }
    files = [
        f for f in sorted(os.listdir(folder))
        if os.path.splitext(f.lower())[1] in extensions
    ]

    results = []
    for i, fname in enumerate(files):
        fpath = os.path.join(folder, fname)

        result = RomScanResult(filepath=fpath, filename=fname)

        try:
            def _prog(pct: float, _fname=fname):
                if on_progress:
                    on_progress(_fname, pct)

            result.crc32 = calc_crc32(fpath, on_progress=_prog)

            # Search all DATs
            for index, dat_name, dat_source in dat_indexes:
                if result.crc32 in index:
                    entry = index[result.crc32]
                    result.matched        = True
                    result.canonical_name = entry["name"]
                    result.rom_name       = entry["rom_name"]
                    result.dat_source     = dat_source
                    break

        except Exception as e:
            result.error = str(e)

        results.append(result)
        if on_result:
            on_result(result)

    return results


# ── Rename ────────────────────────────────────────────────────────────────────

@dataclass
class RenameResult:
    original_path:  str
    new_path:       str
    success:        bool
    error:          str = ""

def rename_roms(
    results:    list[RomScanResult],
    dry_run:    bool = True,
    on_renamed: Callable[[RenameResult], None] | None = None,
) -> list[RenameResult]:
    """
    Rename ROM files to their canonical names from the DAT.

    Args:
        results: List of RomScanResult from scan_folder()
        dry_run: If True, simulate renames without touching files
        on_renamed: Callback called for each rename attempt

    Returns:
        List of RenameResult
    """
    rename_results = []

    for scan in results:
        if not scan.matched:
            continue
        if scan.filename == scan.suggested_filename:
            continue  # already correctly named

        folder    = os.path.dirname(scan.filepath)
        new_path  = os.path.join(folder, scan.suggested_filename)
        r = RenameResult(original_path=scan.filepath, new_path=new_path, success=False)

        try:
            if not dry_run:
                if os.path.exists(new_path) and new_path != scan.filepath:
                    r.error = f"Target already exists: {scan.suggested_filename}"
                else:
                    os.rename(scan.filepath, new_path)
            r.success = True
        except Exception as e:
            r.error = str(e)

        rename_results.append(r)
        if on_renamed:
            on_renamed(r)

    return rename_results

File: core/lib/src/rom_renamer.dart
// OpenROM — Universal ROM Compression Suite
// M5 Dev | GPL v3

import 'dart:io';
import 'dart:typed_data';
import 'package:path/path.dart' as p;
import 'package:xml/xml_events.dart';
import 'config.dart' as config;

final Uint32List _crc32Table = () {
  final table = Uint32List(256);
  for (int i = 0; i < 256; i++) {
    int c = i;
    for (int k = 0; k < 8; k++) {
      if ((c & 1) != 0) {
        c = 0xEDB88320 ^ (c >> 1);
      } else {
        c = c >> 1;
      }
    }
    table[i] = c;
  }
  return table;
}();

int getCrc32(List<int> bytes, [int crc = 0]) {
  int c = crc ^ 0xFFFFFFFF;
  for (int i = 0; i < bytes.length; i++) {
    c = _crc32Table[(c ^ bytes[i]) & 0xFF] ^ (c >> 8);
  }
  return c ^ 0xFFFFFFFF;
}

String getDatsDir() {
  final datsDir = p.join(config.getDirectoryConfig(), 'dats');
  Directory(datsDir).createSync(recursive: true);
  return datsDir;
}

Map<String, dynamic> _readDatHeaderSync(String datPath) {
  String datName = '';
  String datUrl = '';
  int gameCount = 0;
  bool headerFound = false;

  final content = File(datPath).readAsStringSync();
  final events = parseEvents(content);

  String? currentTag;
  for (final event in events) {
    if (event is XmlStartElementEvent) {
      currentTag = event.name;
      if (event.name == 'game') {
        gameCount++;
      }
    } else if (event is XmlTextEvent) {
      if (currentTag == 'name' && datName.isEmpty) {
        datName = event.value.trim();
      } else if (currentTag == 'url' && datUrl.isEmpty) {
        datUrl = event.value.trim();
      }
    } else if (event is XmlEndElementEvent) {
      if (event.name == 'header') {
        headerFound = true;
      }
      currentTag = null;
    }
  }

  if (!headerFound && datName.isEmpty) {
    datName = p.basename(datPath);
  }

  final String source;
  if (datUrl.toLowerCase().contains('no-intro')) {
    source = 'No-Intro';
  } else if (datUrl.toLowerCase().contains('redump')) {
    source = 'Redump';
  } else {
    source = 'Unknown';
  }

  return {
    'name': datName,
    'source': source,
    'game_count': gameCount,
    'url': datUrl,
  };
}

Map<String, dynamic> importDat(String datPath) {
  if (!File(datPath).existsSync()) {
    throw ArgumentError('File not found: $datPath');
  }

  final headerInfo = _readDatHeaderSync(datPath);
  final datName = headerInfo['name'] as String;
  final datUrl = headerInfo['url'] as String;
  final source = headerInfo['source'] as String;
  final gameCount = headerInfo['game_count'] as int;

  final safeName = datName.replaceAll(RegExp(r'[^a-zA-Z0-9\ ._\-\(\)]'), '_');
  final storedPath = p.join(getDatsDir(), '$safeName.dat');
  File(datPath).copySync(storedPath);

  return {
    'name': datName,
    'source': source,
    'system': datName,
    'game_count': gameCount,
    'stored_path': storedPath,
    'url': datUrl,
  };
}

List<Map<String, dynamic>> listDats() {
  final datsDir = getDatsDir();
  final dir = Directory(datsDir);
  final result = <Map<String, dynamic>>[];
  if (!dir.existsSync()) return result;

  final entries = dir.listSync()..sort((a, b) => a.path.compareTo(b.path));
  for (final entry in entries) {
    if (entry is File && entry.path.endsWith('.dat')) {
      try {
        final info = _readDatHeaderSync(entry.path);
        info['stored_path'] = entry.path;
        result.add(info);
      } catch (_) {}
    }
  }
  return result;
}

void removeDat(String storedPath) {
  final file = File(storedPath);
  if (file.existsSync()) {
    file.deleteSync();
  }
}

String calcCrc32(String filepath, {void Function(double pct)? onProgress}) {
  final file = File(filepath);
  final fileSize = file.lengthSync();
  int bytesRead = 0;
  const chunkSize = 1024 * 1024; // 1MB

  final raf = file.openSync(mode: FileMode.read);
  var crc = 0;

  try {
    while (true) {
      final chunk = raf.readSync(chunkSize);
      if (chunk.isEmpty) break;
      crc = getCrc32(chunk, crc);
      bytesRead += chunk.length;
      if (onProgress != null && fileSize > 0) {
        onProgress(bytesRead / fileSize * 100.0);
      }
    }
  } finally {
    raf.closeSync();
  }

  final hex = (crc & 0xFFFFFFFF).toRadixString(16).toLowerCase().padLeft(8, '0');
  return hex;
}

Map<String, Map<String, dynamic>> loadDatIndex(String datPath) {
  final index = <String, Map<String, dynamic>>{};
  final content = File(datPath).readAsStringSync();
  final events = parseEvents(content);

  String currentGameName = '';
  String currentDesc = '';
  String? currentTextTag;

  for (final event in events) {
    if (event is XmlStartElementEvent) {
      if (event.name == 'game') {
        currentGameName = '';
        currentDesc = '';
        for (final attr in event.attributes) {
          if (attr.name == 'name') {
            currentGameName = attr.value;
          }
        }
      } else if (event.name == 'description') {
        currentTextTag = 'description';
      } else if (event.name == 'rom') {
        String crc = '';
        String romName = '';
        int size = 0;
        String md5 = '';
        String sha1 = '';

        for (final attr in event.attributes) {
          if (attr.name == 'crc') {
            crc = attr.value.trim().toLowerCase().padLeft(8, '0');
          } else if (attr.name == 'name') {
            romName = attr.value;
          } else if (attr.name == 'size') {
            size = int.tryParse(attr.value) ?? 0;
          } else if (attr.name == 'md5') {
            md5 = attr.value.trim().toLowerCase();
          } else if (attr.name == 'sha1') {
            sha1 = attr.value.trim().toLowerCase();
          }
        }

        if (crc.isNotEmpty) {
          index[crc] = {
            'name': currentGameName,
            'description': currentDesc.isNotEmpty ? currentDesc : currentGameName,
            'rom_name': romName,
            'size': size,
            'md5': md5,
            'sha1': sha1,
          };
        }
      }
    } else if (event is XmlTextEvent) {
      if (currentTextTag == 'description') {
        currentDesc = event.value.trim();
      }
    } else if (event is XmlEndElementEvent) {
      if (event.name == 'description') {
        currentTextTag = null;
      }
    }
  }

  return index;
}

class RomScanResult {
  final String filepath;
  final String filename;
  String crc32;
  bool matched;
  String canonicalName;
  String romName;
  String datSource;
  String error;

  RomScanResult({
    required this.filepath,
    required this.filename,
    this.crc32 = '',
    this.matched = false,
    this.canonicalName = '',
    this.romName = '',
    this.datSource = '',
    this.error = '',
  });

  String get suggestedFilename {
    if (!matched || romName.isEmpty) return filename;
    return romName;
  }
}

List<RomScanResult> scanFolder(
  String folder,
  List<String> datPaths, {
  void Function(String filename, double percent)? onProgress,
  void Function(RomScanResult result)? onResult,
}) {
  final dir = Directory(folder);
  if (!dir.existsSync()) {
    throw ArgumentError('Folder not found: $folder');
  }

  final datIndexes = <Map<String, dynamic>>[];
  final skippedDats = <Map<String, String>>[];

  for (final dp in datPaths) {
    try {
      final header = _readDatHeaderSync(dp);
      final index = loadDatIndex(dp);
      datIndexes.add({
        'index': index,
        'name': header['name'],
        'source': header['source'],
      });
    } catch (e) {
      skippedDats.add({'name': p.basename(dp), 'reason': e.toString()});
    }
  }

  for (final sk in skippedDats) {
    if (onProgress != null) {
      onProgress('[WARN] Skipped unreadable DAT: ${sk['name']} (${sk['reason']})', 0.0);
    }
  }

  const extensions = {
    '.smc', '.sfc', '.nes', '.gba', '.gb', '.gbc',
    '.n64', '.z64', '.v64', '.ndd',
    '.md', '.gen', '.sms', '.gg', '.32x',
    '.pce', '.iso', '.bin', '.img', '.chd',
    '.ws', '.wsc', '.ngp', '.ngc',
    '.lnx', '.vb', '.vec', '.a26', '.a52', '.j64',
  };

  final entries = dir.listSync();
  final files = <String>[];
  for (final entry in entries) {
    if (entry is File) {
      final fname = p.basename(entry.path);
      final ext = p.extension(fname).toLowerCase();
      if (extensions.contains(ext)) {
        files.add(fname);
      }
    }
  }
  files.sort();

  final results = <RomScanResult>[];
  for (final fname in files) {
    final fpath = p.join(folder, fname);
    final result = RomScanResult(filepath: fpath, filename: fname);

    try {
      result.crc32 = calcCrc32(fpath, onProgress: (pct) {
        if (onProgress != null) onProgress(fname, pct);
      });

      for (final datEntry in datIndexes) {
        final index = datEntry['index'] as Map<String, Map<String, dynamic>>;
        final datSource = datEntry['source'] as String;

        if (index.containsKey(result.crc32)) {
          final item = index[result.crc32]!;
          result.matched = true;
          result.canonicalName = item['name'] as String;
          result.romName = item['rom_name'] as String;
          result.datSource = datSource;
          break;
        }
      }
    } catch (e) {
      result.error = e.toString();
    }

    results.add(result);
    if (onResult != null) {
      onResult(result);
    }
  }

  return results;
}

class RenameResult {
  final String originalPath;
  final String newPath;
  bool success;
  String error;

  RenameResult({
    required this.originalPath,
    required this.newPath,
    required this.success,
    this.error = '',
  });
}

List<RenameResult> renameRoms(
  List<RomScanResult> results, {
  bool dryRun = true,
  void Function(RenameResult r)? onRenamed,
}) {
  final renameResults = <RenameResult>[];

  for (final scan in results) {
    if (!scan.matched) continue;
    if (scan.filename == scan.suggestedFilename) continue;

    final folder = p.dirname(scan.filepath);
    final newPath = p.join(folder, scan.suggestedFilename);
    final r = RenameResult(
      originalPath: scan.filepath,
      newPath: newPath,
      success: false,
    );

    try {
      if (!dryRun) {
        if (File(newPath).existsSync() &&
            p.canonicalize(newPath) != p.canonicalize(scan.filepath)) {
          r.error = 'Target already exists: ${scan.suggestedFilename}';
        } else {
          File(scan.filepath).renameSync(newPath);
          r.success = true;
        }
      } else {
        r.success = true;
      }
    } catch (e) {
      r.error = e.toString();
    }

    renameResults.add(r);
    if (onRenamed != null) {
      onRenamed(r);
    }
  }

  return renameResults;
}
